src/main/scala/mesosphere/marathon/core/check/Check.scala

```diff
@@ -0,0 +1,254 @@
+package mesosphere.marathon
+package core.check
+
+import com.wix.accord._
+
+import mesosphere.marathon.Protos.CheckDefinition.Protocol
+import mesosphere.marathon.state._
+import org.apache.mesos.{Protos => MesosProtos}
+
+import scala.concurrent.duration._
+
+/**
+  * Check is a Type to help with type conversions to and from different protos.
+  * It mirrors closely to HealthCheck which does the same thing.
+  * Check is the most abstract Check.   The appDef works with MesosChecks.
+  *
+  * toProto takes this data structure into the CheckDefinition proto defined in Marathon used for storage.
+  * toMesos takes this data structure into Mesos CheckInfo for working with TaskBuilder
+  * Conversations to and from raml is in CheckConversion class.
+  */
+sealed trait Check {
+  def delay: FiniteDuration
+  def interval: FiniteDuration
+  def timeout: FiniteDuration
+  def toProto: Protos.CheckDefinition
+
+  protected def protoBuilder: Protos.CheckDefinition.Builder =
+    Protos.CheckDefinition.newBuilder
+      .setIntervalSeconds(this.interval.toSeconds.toInt)
+      .setTimeoutSeconds(this.timeout.toSeconds.toInt)
+      .setDelaySeconds(this.delay.toSeconds.toInt)
+}
+
+sealed trait PortReference extends Product with Serializable {
+  def apply(assignments: Seq[PortAssignment]): PortAssignment
+  def buildProto(builder: Protos.CheckDefinition.Builder): Unit
+}
+
+object PortReference {
+  case class ByIndex(value: Int) extends PortReference {
+    override def apply(assignments: Seq[PortAssignment]): PortAssignment =
+      assignments(value)
+    override def buildProto(builder: Protos.CheckDefinition.Builder): Unit =
+      builder.setPortIndex(value)
+  }
+
+  def apply(value: Int): PortReference = ByIndex(value)
+
+  def fromProto(pb: Protos.CheckDefinition): Option[PortReference] =
+    if (pb.hasPortIndex) Some(ByIndex(pb.getPortIndex))
+    else if (!pb.hasPort) Some(ByIndex(0))
+    else None
+}
+
+sealed trait CheckWithPort extends Check {
+  def portIndex: Option[PortReference]
+  def port: Option[Int]
+}
+
+object CheckWithPort {
+  val DefaultPortIndex = None
+  val DefaultPort = None
+
+  import mesosphere.marathon.api.v2.Validation.isTrue
+
+  implicit val Validator: Validator[CheckWithPort] =
+    isTrue("Check must specify either a port or a portIndex") { hc =>
+      hc.portIndex.isDefined ^ hc.port.isDefined
```

Comment:
Fancy! Would love to see unit test for this

Reply:
I guess it would be `!=` then:
```hc.portIndex.isDefined != hc.port.isDefined```

Reply:
This is an XOR on booleans which is exactly what is meant 
`||` (or) doesn't work without additional logic 
XOR truth table
INPUT | OUTPUT
-- | --
A | B | A XOR B
F | F | F
F | T | T
T | F | T
T | T | F